include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1270 @@
+        values = values_double;
+      }
+
+      static void
+      jacobian (const unsigned int             &active_tape_index,
+                const unsigned int             &n_dependent_variables,
+                const std::vector<scalar_type> &independent_variables,
+                FullMatrix<scalar_type>        &jacobian)
+      {
+        FullMatrix<double> jacobian_double (jacobian.m(), jacobian.n());
+        TapedDrivers<ADNumberType,double>::jacobian(
+          active_tape_index,
+          n_dependent_variables,
+          vector_float_to_double(independent_variables),
+          jacobian_double);
+        jacobian = jacobian_double;
+      }
+    };
+
+
+    // -------------   TapelessDrivers   -------------
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapelessDrivers<ADNumberType,ScalarType,T>::initialize (
+      const unsigned int &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    ScalarType
+    TapelessDrivers<ADNumberType,ScalarType,T>::value (
+      const std::vector<ADNumberType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+      return ScalarType(0.0);
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapelessDrivers<ADNumberType,ScalarType,T>::gradient (
+      const std::vector<ADNumberType> &,
+      const std::vector<ADNumberType> &,
+      Vector<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapelessDrivers<ADNumberType,ScalarType,T>::hessian (
+      const std::vector<ADNumberType> &,
+      const std::vector<ADNumberType> &,
+      FullMatrix<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapelessDrivers<ADNumberType,ScalarType,T>::values (
+      const std::vector<ADNumberType> &,
+      Vector<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapelessDrivers<ADNumberType,ScalarType,T>::jacobian (
+      const std::vector<ADNumberType> &,
+      const std::vector<ADNumberType> &,
+      FullMatrix<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    namespace internal
+    {
+
+      // A dummy function to define the active dependent variable when using
+      // reverse-mode AD.
+      template<typename ADNumberType>
+      static typename std::enable_if<
+      !is_sacado_rad_number<ADNumberType>::value
+      >::type
+      reverse_mode_dependent_variable_activation (ADNumberType &)
+      {
+
+      }
+
+#ifdef DEAL_II_WITH_TRILINOS
```

Comment:
DEAL_II_WITH_ADOLC?

Reply:
No, this is enabled if `is_sacado_rad_number` (so definitely Trilinos)

Reply:
ah, we don't have a separate `DEAL_II_WITH_SACADO`, meaning that Sacado is required dependency?

Reply:
Yes, its [been this way](https://github.com/dealii/dealii/blob/master/cmake/configure/configure_2_trilinos.cmake#L45) for a long time. Sometime in the future I'd like to modularise our Trilinos configuration.

Reply:
Thanks to #6906, I can now add this finer-grained check :-)